File: lib/flutter/api/app_exception.dart
class AppException implements Exception {
  final String? _message;
  final String? _prefix;

  AppException([this._prefix, this._message]);

  @override
  String toString() {
    return '$_prefix$_message';
  }
}

class UnSupportImageTypeError extends AppException {
  UnSupportImageTypeError([String? message])
      : super("Invalid Type Request: ", message);
}

class BadRequestError extends AppException {
  BadRequestError([String? message]) : super("Bad Request: ", message);
}

class NoInternetError extends AppException {
  NoInternetError([String? message]) : super("No Internet: ", message);
}

class UnknownError extends AppException {
  UnknownError([String? message]) : super("Unknown Error Occur");
}

File: plugin/download/android/src/main/kotlin/com/example/perfect_corp_homework/download/exception/AppException.kt
package com.example.perfect_corp_homework.download.exception

import com.example.perfect_corp_homework.download.kBadRequestErrorMessage
import com.example.perfect_corp_homework.download.kNoInternetErrorMessage
import com.example.perfect_corp_homework.download.kUnSupportMediaTypeErrorMessage

open class AppException(): Exception()

class UnSupportImageTypeError(): AppException() {
    override val message: String
        get() = kUnSupportMediaTypeErrorMessage
}

class BadRequestError(): AppException() {
    override val message: String
        get() = kBadRequestErrorMessage
}

class NoInternetError(): AppException() {
    override val message: String
        get() = kNoInternetErrorMessage
}

class UnknownError(private var error: Exception): AppException() {
    override val message: String
            get() = "UnExpected Error: ${error.message}"
    override val cause: Throwable
        get() = error
}

class JsonSerializationError(private var error: Exception): AppException() {
    override val message: String
        get() = "JsonSerializationError: ${error.message}"
}
